### PR TITLE
Allow to INSERT or DELETE on tables with a row filter

### PR DESCRIPTION
**Pull request by kokosing**
Allow to INSERT or DELETE on tables with a row filter

This change makes it possible to execute INSERT or DELETE
on tables with row filters.

We consider table with a row filter as a form of a dynamic view
with a CHECK OPTION.

- INSERT, if the row that's inserted does not satisfy the
definition of the view, the query is expected to fail as CHECK OPTION
is in effect.

- DELETE can operate only on visible (filtered) rows.

This change does not affect UPDATE operation. This is still not
supported whenever table has a row filter.

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 1]
```diff
@@ -341,7 +341,7 @@ public Scope analyze(Node node, Optional<Scope> outerQueryScope)
                 .process(node, Optional.empty());
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:341]: "Silence IDE warnings" is not a good explanation. What's the actual logical change being done? Please describe it in those terms and break up the commit into smaller ones if the changes are unrelated (e.g., constrain method visibility, replace lambda expressions with method references, etc)

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 2]
```diff
@@ -1866,7 +1866,13 @@ protected Scope visitAliasedRelation(AliasedRelation relation, Optional<Scope> s
             if (relation.getColumnNames() != null) {
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:1870]: These seem unnecessary. We don't use hard-wrapping. If your IDE is changing them upon reformatting, you may want to update your IDE settings.

The general rule is: anything that is relevant to make sense of the code should fit within view in a typical screen (e..g, a laptop screen), but non-essential things can extend beyond -- for example, the elements making up an exception message, which don't affect the flow of code or logic.

File: core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java [hunk 3]
```diff
@@ -658,7 +658,8 @@ private Expression noTruncationCast(Expression expression, Type fromType, Type t
 
     private RelationPlan createDeletePlan(Analysis analysis, Delete node)
     {
-        DeleteNode deleteNode = new QueryPlanner(analysis, symbolAllocator, idAllocator, buildLambdaDeclarationToSymbolMap(analysis, symbolAllocator), metadata, Optional.empty(), session, ImmutableMap.of())
+        Map<NodeRef<LambdaArgumentDeclaration>, Symbol> lambdaDeclarationToSymbolMap = buildLambdaDeclarationToSymbolMap(analysis, symbolAllocator);
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java:695]: These seem unnecessary. Moreover, the extra variable (with the overly verbose name) makes it harder to read the code. If the length of the line is an issue, it would more readable to format this as:

```java
DeleteNode deleteNode = new QueryPlanner(
        analysis, 
        symbolAllocator, 
        idAllocator, 
        buildLambdaDeclarationToSymbolMap(analysis, symbolAllocator), 
        metadata, 
        Optional.empty(), 
        session, 
        ImmutableMap.of())
```

**Comment by martint** [comment]
Row filters are a form of dynamic view. With this change, we're implementing a subset of what the spec describes as `UPDATABLE VIEW`s, so we should make sure the semantics match what the spec requires.

**Comment by kokosing** [comment]
> Row filters are a form of dynamic view. With this change, we're implementing a subset of what the spec describes as UPDATABLE VIEWs, so we should make sure the semantics match what the spec requires.

@martint Would you please share spec fragment that would explain the semantics of it?

**Comment by kokosing** [comment]
Extracted https://github.com/trinodb/trino/pull/8898 and https://github.com/trinodb/trino/pull/8897. I am going to drop the commit with indentation changes.

**Comment by findepi** [comment]
> In case of INSERT when row is matching a row filter then query will
fail.

the conditions imposed by a row filter are not considered secret, right?

i know we expose them in EXPLAIN today, but in the future we could choose not to (eg even by disallowing EXPLAIN in presence of row filters, or by masking them out).

Allowing INSERT but only if it satisfies row filters will limit our options.

> In case of DELETE and UPDATE rows that are not matching the row filters
> are not read at all and so they are not modified or deleted.

it's not intuitive that `UPDATE my_table SET injection_date = CAST(injection_timestamp AS date)` will leave the `injection_date` null for some rows.
shouldn't we fail instead?

but then, failing would reveal that there are some rows that match delete/update criteria which cannot be seen, which would be security issue.

therefore, I am in favor of NOT enabling this at all.


Maybe, as an alternative, we should consider supporting INSERT, UPDATE and DELETE for views? It would have more well-defined semantics, i guess.

cc @losipiuk @hashhar

**Comment by kokosing** [comment]
>> In case of DELETE and UPDATE rows that are not matching the row filters
are not read at all and so they are not modified or deleted.

> it's not intuitive that UPDATE my_table SET injection_date = CAST(injection_timestamp AS date) will leave the injection_date null for some rows.
shouldn't we fail instead?

I don't think we will leave any nulls. The statement will be "translated" into `UPDATE my_table SET injection_date = CAST(injection_timestamp AS date) WHERE <row filter>`. So value of `injection_date` will be unchanged for rows not matching filter.

> Maybe, as an alternative, we should consider supporting INSERT, UPDATE and DELETE for views? It would have more well-defined semantics, i guess.

Actually, this is actually it.  As @martint suggested in https://github.com/trinodb/trino/pull/8856#issuecomment-899883351, it is a form of dynamic view. But I am not yet sure about their semantics yet.

**Comment by findepi** [comment]
> I don't think we will leave any nulls. 

i misspoke. updated above.

> > Maybe, as an alternative, we should consider supporting INSERT, UPDATE and DELETE for views? It would have more well-defined semantics, i guess.
> 
> Actually, this is actually it. As @martint suggested in [#8856 (comment)](https://github.com/trinodb/trino/pull/8856#issuecomment-899883351), it is a form of dynamic view.

It is a view from "mechanics" perspective, but it's not a view from user perspective.
I.e. a view is explicitly not a table, so updating a view may or may not update all rows in the underlying table(s).

Filtered table is still a table, and filter is implicit. User may not know the filter is applied, and all users refer to same table by the same name, but see different data.

**Comment by kokosing** [comment]
@martint Any update about the semantic of how row filters should be applied for INSERT, DELETE and UPDATE?

**Comment by kokosing** [comment]
@martint ping

**Comment by martint** [comment]
> @martint Any update about the semantic of how row filters should be applied for INSERT, DELETE and UPDATE?

I'm going over the spec, but it's large and complicated. My understanding so far is:

* For UPDATE, any rows that are "visible" in the result of the view are candidates for update. The update operation logically follows these steps:
    1. The candidate row for UPDATE is identified
    2. A copy of the candidate row is created and the changes from the SET assignments are applied to it
    3. The original row is replaced with the new row
* For INSERT:
    1. For each row in the source, a new candidate row is created with any columns that have DEFAULT set to the corresponding default value.
    2. The columns declared in the INSERT clause are copied from the source row to the candidate row
    2. The candidate row is inserted into the target table

Unless the view is declared WITH CHECK OPTION, no verification is performed to make sure the updated row satisfies the definition of the view. We don't support this in the view definition syntax, yet.

**Comment by kokosing** [comment]
As I understand the current implementation for UPDATE matches the specification. We only update rows that are visible to user.

But for INSERT current implementation fails when trying to insert row that is not matching the row filter. Like it would be `WITH CHECK OPTION`. So to make it work according to spec, I should simply ignore rows that do not match the row filter. In other words, I should put the row filter on the source side of the insert.

@martint Please double check if what I am saying is correct. If so, I will update the PR.

CC: @findepi

**Comment by martint** [comment]
UPDATE can operate on visible rows, of course. However, if the result of applying the UPDATE operation would result in the row not being visible, the query is expected to fail only if WITH CHECK OPTION is declared for the view.

For INSERT, if the row that's inserted does't satisfy the definition of the view, the query should fail if WITH CHECK OPTION is declared for the view. Otherwise, the row should be inserted as is without any further checking.

**Comment by kokosing** [comment]
The last question. Should we treat table with a row filter as a view WITH CHECK OPTION or as a view without CHECK OPTION?

**Comment by findepi** [comment]
I don't imagine any practical use of row-filtered update _without_ `CHECK OPTION`.

**Comment by kokosing** [comment]
Updated the PR to follow the spec.

**Comment by kokosing** [comment]
@martint Can you please take a look?

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 4]
```diff
@@ -424,10 +424,6 @@ protected Scope visitInsert(Insert insert, Optional<Scope> scope)
             }
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:424]: > In case of INSERT when row is matching a row filter then query will
fail.

According to the SQL spec, it's the other way around. If the row *does not* match the row filter, then the query should fail.

> In case of DELETE and UPDATE rows that are not matching the row filters
are not read at all and so they are not modified or deleted.

Similarly, for UPDATE, if the result of applying the update is a row that does not match the filter, the query should fail. For instance, if the row filter says `WHERE x = 1`,

given a `row(x) == (1)`, the following query should fail:

```
UPDATE t SET x = 2
```

The row qualifies for consideration for the update because it matches the row filter, but the resulting row after the update is applied (i.e., `(2)`) no longer satisfies the filter.

Comment (kokosing) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:424]: I am sorry, my bad. I haven't updated the commit message. It is like was initially. Fixed.

Comment (kokosing) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:424]: @martint ping

**Comment by kokosing** [comment]
@martint @findepi Please take a look. Now I obey `CHECK OPTION`. However I was not able to implement this correctly for UPDATE, it is more complex that for INSERT. So I decided to leave it for another PR.

**Review by Praveen2112 — APPROVED** [review]

**Comment by losipiuk** [comment]
Update commit message. It starts with  "This change makes it possible to execute INSERT, UPDATE or DELETE"
